encoder/vaapiencoder_vp9.cpp

```diff
@@ -0,0 +1,323 @@
+                                          Yami_PTR videoEncParams)
+{
+    YamiStatus status = YAMI_SUCCESS;
+    FUNC_ENTER();
+    if (!videoEncParams)
+        return YAMI_INVALID_PARAM;
+
+    switch (type) {
+    default:
+        status = VaapiEncoderBase::setParameters(type, videoEncParams);
+        break;
+    }
+    return status;
+}
+
+YamiStatus VaapiEncoderVP9::getParameters(VideoParamConfigType type,
+                                          Yami_PTR videoEncParams)
+{
+    FUNC_ENTER();
+    if (!videoEncParams)
+        return YAMI_INVALID_PARAM;
+
+    return VaapiEncoderBase::getParameters(type, videoEncParams);
+}
+
+YamiStatus VaapiEncoderVP9::doEncode(const SurfacePtr& surface,
+                                     uint64_t timeStamp, bool forceKeyFrame)
+{
+    YamiStatus ret;
+    if (!surface)
+        return YAMI_INVALID_PARAM;
+
+    PicturePtr picture(new VaapiEncPictureVP9(m_context, surface, timeStamp));
+
+    m_frameCount %= keyFramePeriod();
+    picture->m_type = (m_frameCount ? VAAPI_PICTURE_P : VAAPI_PICTURE_I);
+    m_frameCount++;
+
+    CodedBufferPtr codedBuffer
+        = VaapiCodedBuffer::create(m_context, m_maxCodedbufSize);
+    if (!codedBuffer)
+        return YAMI_OUT_MEMORY;
+    picture->m_codedBuffer = codedBuffer;
+    codedBuffer->setFlag(ENCODE_BUFFERFLAG_ENDOFFRAME);
+    if (picture->m_type == VAAPI_PICTURE_I) {
+        codedBuffer->setFlag(ENCODE_BUFFERFLAG_SYNCFRAME);
+    }
+    ret = encodePicture(picture);
+    if (ret != YAMI_SUCCESS) {
+        return ret;
+    }
+    output(picture);
+    return YAMI_SUCCESS;
+}
+
+bool VaapiEncoderVP9::fill(VAEncSequenceParameterBufferVP9* seqParam) const
+{
+    seqParam->max_frame_width = kMaxWidth;
+    seqParam->max_frame_height = kMaxHeight;
+
+    seqParam->bits_per_second = bitRate();
+    seqParam->intra_period = intraPeriod();
+    seqParam->kf_min_dist = 1;
+    seqParam->kf_max_dist = intraPeriod();
+
+    return true;
+}
+
+// Fills in VA picture parameter buffer
+bool VaapiEncoderVP9::fill(VAEncPictureParameterBufferVP9* picParam,
+                           const PicturePtr& picture,
+                           const SurfacePtr& surface) const
+{
+    picParam->reconstructed_frame = surface->getID();
+    picParam->coded_buf = picture->getCodedBufferID();
+
+    if (picture->m_type == VAAPI_PICTURE_I) {
+        memset(picParam->reference_frames, 0xFF,
+               sizeof(picParam->reference_frames));
+    }
+    else {
+        picParam->refresh_frame_flags = 0x01; //refresh last frame
+        picParam->pic_flags.bits.frame_type = kInterFrame;
+
+        ReferenceQueue::const_iterator it = m_reference.begin();
+
+        for (uint32_t i = 0; it != m_reference.end(); ++it, i++)
+            picParam->reference_frames[i] = (*it)->getID();
+
+        it = m_reference.begin();
+
+        // last/golden/alt is used as reference frame. L0 forward
+        picParam->ref_flags.bits.ref_frame_ctrl_l0 = 0x7;
```

Comment:
I see ref_frame_ctrl_l1 in define too. When to use l1?  Is it improve image quality if we use l0 and l1 at same time.